util/election/runner.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package election
+
+import (
+	"context"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/util"
+)
+
+// Config is a set of parameters for a continuous master election process.
+type Config struct {
+	// PreElectionPause is the maximum interval to wait before starting election.
+	PreElectionPause time.Duration
+	// MasterCheckInterval is the interval between checks that the instance still
+	// holds mastership.
+	MasterCheckInterval time.Duration
+	// MasterTTL is the time after which the instance will resign if encountered
+	// no successful IsMaster calls.
+	MasterTTL time.Duration
+}
+
+// Run contains mastership status of a single instance during a single run of
+// being master.
+type Run struct {
+	// Start is the time when the instance became the master as measured by the
+	// corresponding Runner's TimeSource.
+	Start time.Time
+	// Ctx is the context which is active while the instance believes to be the
+	// master. As soon as Ctx is canceled, the instance attempts to resign, and
+	// stops monitoring its status.
+	Ctx context.Context
```

Comment:
General advice is not to store contexts in structs -- do you need to?

Reply:
Could you link the rationale of this advice?

My impression is that if something has multiple contexts (not necessarily parent-child related ones), it is okay to store some in structs because passing the whole bunch through a method interface would be painful.

You can see such examples in `etcd` library:
- election's `Campaign` method [uses](https://github.com/coreos/etcd/blob/e2a6489c1ee001d24a3c298f04ee0031f0484631/clientv3/concurrency/election.go#L89) `client`'s context as a back-up to be able to resign despite `ctx` being canceled
- [ctx/cancel](https://github.com/coreos/etcd/blob/e2a6489c1ee001d24a3c298f04ee0031f0484631/clientv3/lease.go#L159:2) stored in lease maintaining object

Not sure how much the examples above justify my case here. Will see if I can simplify this further.

Reply:
https://github.com/golang/go/wiki/CodeReviewComments#contexts has the original advice, but it's a bit lacking in *why*.